general/mem_manager.hpp

```diff
@@ -394,158 +434,254 @@ class Memory
+   void *GetHostPtr(const void *d_ptr, size_t bytes, bool copy_data);
+
+   /// Return the corresponding host pointer of alias_ptr,
+   /// allocating and moving the data if needed
+   void *GetAliasHostPtr(const void *alias_ptr, size_t bytes, bool copy_data);
 
 public:
    MemoryManager();
    ~MemoryManager();
 
+   /// Configure the Memory manager with given default host and device types
+   /// This method will be called when configuring a device.
+   void Configure(const MemoryType h_mt, const MemoryType d_mt);
+
+#ifdef MFEM_USE_UMPIRE
+   /// Set the host and device UMpire allocator names
+   void SetUmpireAllocatorNames(const char *h_name, const char *d_name);
+   const char *GetUmpireAllocatorHostName() { return h_umpire_name; }
+   const char *GetUmpireAllocatorDeviceName() { return d_umpire_name; }
+#endif
+
+   /// Free all the device memories
    void Destroy();
 
-   /// Return true if a global memory manager instance exists
-   static bool Exists() { return exists; }
+   /// Return true if the pointer is known by the memory manager
+   bool IsKnown(const void *h_ptr) { return IsKnown_(h_ptr); }
+
+   /// Return true if the pointer is known by the memory manager as an alias
+   bool IsAlias(const void *h_ptr) { return IsAlias_(h_ptr); }
 
-   /// Check if pointer has been registered in the memory manager
-   void RegisterCheck(void *ptr);
+   /// Check if the host pointer has been registered in the memory manager
+   void RegisterCheck(void *h_ptr);
 
    /// Prints all pointers known by the memory manager
    void PrintPtrs(void);
+
+   /// Prints all aliases known by the memory manager
+   void PrintAliases(void);
+
+   MemoryType GetHostMemoryType() { return host_mem_type; }
+   MemoryType GetDeviceMemoryType() { return device_mem_type; }
 };
 
 
 // Inline methods
 
+template <typename T>
+inline void Memory<T>::Reset()
+{
+   h_ptr = NULL;
+   h_mt = MemoryManager::host_mem_type;
+   capacity = 0;
+   flags = 0;
+}
+
+template <typename T>
+inline void Memory<T>::Reset(MemoryType mt)
+{
+   h_ptr = NULL;
+   h_mt = mt;
+   capacity = 0;
+   flags = 0;
+}
+
+template <typename T>
+inline void Memory<T>::New(int size)
+{
+   capacity = size;
+   flags = OWNS_HOST | VALID_HOST;
+   h_mt = MemoryManager::host_mem_type;
+   h_ptr = (h_mt == MemoryType::HOST) ? new T[size] :
+           (T*)MemoryManager::New_(nullptr, size*sizeof(T), h_mt, flags);
+   MemoryManager::CheckHostMemoryType_(h_mt, h_ptr);
+}
+
 template <typename T>
 inline void Memory<T>::New(int size, MemoryType mt)
 {
-   if (mt == MemoryType::HOST)
-   {
-      New(size);
-   }
-   else
+   capacity = size;
+   if (mt == MemoryType::HOST) { flags = OWNS_HOST | VALID_HOST; }
+   h_mt = IsHostMemory(mt) ? mt : MemoryManager::GetDualMemoryType_(mt);
+   T *h_tmp = (h_mt == MemoryType::HOST) ? new T[size] : nullptr;
+   h_ptr = (mt == MemoryType::HOST) ? h_tmp:
+           (T*)MemoryManager::New_(h_tmp, size*sizeof(T), mt, flags);
```

Comment:
Here, the check `(mt == MemoryType::HOST)` is done three times -- it seems natural to reorganize the code a little so the check is done only once.

Reply:
Two for `mt` plus one for `h_mt`, but I'll factorize the first one, ok.

Reply:
Simplified, `done`